### DISCUSSION TITLE
Specific data(context?) per state?

### DISCUSSION THREAD

**Discussion by lrxw** [statelyai/xstate#3258] [Q&A]
HI, I just started looking into xstate and if its feasible to use in our use case. Having all the states, events and valid transitions defined, looks really promising.
In our application we have a form, doing calculations and depending on the results, enable buttons, deactivate them, showing results or error messages.
I wonder if there is a way to have a context object (and type) per state? For example, if I have a state `error`, I want to be sure, there is an error object. If I have a piece of code like:

```
if (current.matches('error')) {
    return <ErrorComponent error={current.context.error} />
}
```

There is no way Typescript knows, that the error field of the context is present when it matches the error state. Currently we check both state and object:

```
if (current.matches('error') && current.context.error) {
    return <ErrorComponent error={current.context.error} />
}
```

**Comment by Andarist** [comment]
> There is no way Typescript knows, that the error field of the context is present when it matches the error state.

It's because with the current TS possibilities and our API this is not possible. There are some open PRs to TypeScript that might bring us closer to this - but it's a complex problem and it might take us time to get us there even with those PRs merged in and released.

**Comment by lrxw** [comment]
Which PRs are you talking about?
I tried to define my own types based on the types from xstate. The typing looks very complicated and I'm probably not the best to look into this. I tried it that way (Disclaimer, this is very rudimentary and just a first step):

```
type MyStateSchema<TC> = StateSchema & {
  context: TC
  states?: {
    [key: string]: StateSchema<TC>
  }
}

type InitialContext = {}

export interface MyMachineConfig
  extends MachineConfig<any, MyStateSchema<any>, any> {
  states: {
    [AppState.INITIAL]: MyStateSchema<InitialContext>
    [AppState.CALCULATING]: MyStateSchema<null>
    [AppState.UNSAVED]: MyStateSchema<InitialContext>
  }
}
```
but in the end, the state machine is missing a functionality to send the state specific data with an event and to store it in the state.
